Feature/api 2.create question - 텍스트에서 질문 추출

silofn523 opened the pull request:
## ⚙️ 작업 내용

> Gemini API와 연결해서 텍스트에서 질문 추출

## 🔎작업 상세 내용

- Gemini-API연동하여 텍스트 기반으로 질문 추출하는 API개발
- 커스텀 예외처리 (Gemini err, DB err)
- DDD를 적용시켜 역할별 분리
- Redis를 이용하여 캐싱 (같은 텍스트 값으로 2번 이상 요청 보낼시)
- QuestionSet 엔티티에 사용자가 입력한 text값 저장, Question엔티티와 외래키로 연결 (어떤 텍스트를 기반으로 질문들이 생성됐는지 파악하기위해)
- Question 엔티티에 질문 리스트 저장

## 📸 스크린샷 

<img width="694" alt="스크린샷 2025-04-10 오후 9 32 23" src="https://github.com/user-attachments/assets/952780e0-1133-4e30-8d28-83e76ce23013" />

## ⚠️ 주의 사항 & 중점적으로 봐야 할 부분
- 깃허브 .env파일만 지우려다 피처 브렌치가 복사가 되면서 이상한 오류 때문에 시간이 오래걸림&잘 고쳐진것도 모르겠음... 
## 💫참고할만한 자료(선택)

silofn523 committed:
API-2 질문 추출 더미 데이터값으로 엔드포인트 구축

silofn523 committed:
Gemini-API로 텍스트에서 질문 추출
변경 사항
- Gemini-API연동하여 텍스트 기반으로 질문 추출하는 API개발
- 커스텀 예외처리 (Gemini err, DB err)
- DDD를 적용시켜 역할별 분리
- Redis를 이용하여 캐싱 (같은 텍스트 값으로 2번 이상 요청 보낼시)
- QuestionSet 엔티티에 사용자가 입력한 text값 저장, Question엔티티와 외래키로 연결 (어떤 텍스트를 기반으로 질문들이 생성됐는지 파악하기위해)
- Question 엔티티에 질문 리스트 저장

기타 참고 사항
- 관련 이슈: 2

silofn523 committed:
중복 파일 삭제

oakkim commented on src/question/integrations/gemini-api.service.ts:
```diff
@@ -0,0 +1,98 @@
+    private readonly configService: ConfigService
+  ) {
+    this.genAI = new GoogleGenerativeAI(process.env.GEMINI_API_KEY)
+    this.model = this.genAI.getGenerativeModel({ model: 'gemini-1.5-flash' })
```
2.0-flash 써도 될듯?

oakkim commented on src/question/integrations/gemini-api.service.ts:
```diff
@@ -0,0 +1,98 @@
+import { createHash } from 'crypto'
+
+@Injectable()
+export class GeminiApiService {
```
요 서비스의 이름을 `GenerateQuestionService`, `QuestionGenerationService` 등으로 바꾸면 어떨까? 질문을 생성하는 역할 하나만 부여하는거지. 나중에 질답을 하는 용의 서비스가 따로 추가될 수 있을 것 같네.

oakkim replied:
좀 더 고오급지게 하려면, interface를 활용하기 딱 좋을 것 같아. 꼭 Gemini를 쓰지 않고, 다른 모델을 쓰게되면 이 클래스를 사용하는 다른 쪽까지 바꿔야하는 강한 결합이 일어나겠지? 그래서 인터페이스를 써보면 좋은 예제 같아. (안해도됨)

oakkim commented on src/question/integrations/gemini-api.service.ts:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  /**
+   * 동일한 텍스트에 대해 반복적으로 API 호출 방지 - 캐싱
```
캐싱 굿

oakkim replied:
근데 갑자기 생각났는데, 만약 사용자가 다른 질문을 받아보고 싶다면 어떻게 해야할까? 같은 텍스트라서 캐싱에 무조건 걸릴 수 밖에 없겠네. 해당 클래스는 딱 LLM 모델을 통한 질문 생성만 하고, 이 클래스를 사용하는 쪽에서 캐싱을 할지말지 제어하는게 어떨까?

oakkim reviewed (approved):
굿 수고했으. 내 코멘트 보고 적용할지말지는 판단해보면 좋을 것 같아. 물어봐도 좋고.

silofn523 committed:
기술별&프로젝트별 질문 생성 및 프롬프트 수정

변경 사항
- 기술별 질문 엔티티 설계 및 생성
- 질문 생성시에 기술별 질문과 프로젝트 질문으로 생성
- 기술별 질문, 프로젝트별 질문 response DTO 생성
- 프롬프트 수정

기타 참고 사항
- 관련 이슈: 2
